### DISCUSSION TITLE
How to enrich matrix output to be able detect uturns on combinations of matrix elements?

### DISCUSSION THREAD

**Discussion by radekvermirovsky** [valhalla/valhalla#4814] [Ideas]
Hi, we would like to have the ability to address U-turn penalties in TSP and VRP heuristics. To make this possible, we need to know if a two elements from the matrix creates U-turn or not. As a simple solution, we thought that the matrix could return the azimuth for each element in two values. One would be the azimuth of the departure from the source and the other would be the azimuth of the arrival at the target. If we had these two pieces of information for each element, we could determine that if the azimuth at the target of the first element is exactly opposite to the azimuth at the source of the element that follows, we would know it’s a U-turn. This is mainly an initial thought. Do you have any better solutions that could be introduced to the matrix transitions and subsequently evaluate whether a U-turn occurs in various combinations of transitions? At start we were thinking about looking at indices of edges, but than we realize this will fail on trivial routes heavily.  So azimuth sounds like pretty reliable solution. Also its probably cheap to compute it or its maybe already there. Thanks for your insights.

**Comment by chrstnbwnkl** [comment]
Valhalla stores headings on nodes, so it would be easy to get those for the source/target edges when we do the recosting (that's when we construct the actual path). Why do you think this would fail on trivial routes though? Or at least I don't see how the azimuth strategy would succeed but the edge ID would fail but maybe I'm forgetting something :smile: In any case I think it's possible, maybe even via edge indices (return the ID of the start edge and the ID of the end edge's opposite?)

**Reply by radekvermirovsky** [reply]
In the case of an ID, when I have two calculated elements (matrix pairs) and both are on the same edge (both are trivial routes), I would get the same edge ID at both the start and the end for each element. If I then wanted to use the TSP algorithm to determine whether combining these two elements results in a U-turn, I wouldn't be able to figure it out because all I would have for comparison would be the same edge ID four times. However, if edges in different directions on the same road had different IDs (as I would expect), this approach wouldn't work.

If I had the heading (I previously mentioned azimuth, which was incorrect—I meant heading), I could compare that instead. What I actually need is the heading at the points where I start and end on the edge. If each edge is always defined by just two points and a straight line, then the heading of the node before my start or beyond my endpoint would match the heading at the point where I start or finish the element. In this case, the headings of the respective nodes would work fine, and it would function correctly even if both elements are computed on the same edge.

If I combine these elements in the same direction, the heading at the end of the first element will be the same as the heading at the start of the second element, and no U-turn will be detected. However, if the second element is on the same edge but in the reverse direction, the heading at the end of the first element will differ by 180 degrees, and a U-turn will be detected.

What I don’t understand is how the heading is defined at a node. When I have a single point with multiple edges converging into it, does each edge have its own node, or is there just one node for all of them? If there is only one node for all edges, is the heading stored multiple times at that node, once for each edge?

**Reply by chrstnbwnkl** [reply]
Regarding your last point: headings are stored for each outgoing edge from a node at 2 degree precision. It can take up to 64 bits, with each heading taking 8 bits, i.e. headings can be stored for up to 8 outgoing edges.

**Reply by chrstnbwnkl** [reply]
And yes you're right of course, I wasn't thinking about long and winding edges :smile: in that case though the edge heading we store on the node is not helpful. But I think the heading you are looking for is actually already computed during the location search, and those results should be on the Protobuf response, so maybe nothing needs to be done on Valhalla's side (unless you really want it as JSON).
